test/src/com/xilinx/rapidwright/eco/TestECOTools.java

```diff
@@ -0,0 +1,438 @@
+        EDIFNetlist netlist = design.getNetlist();
+        Map<Net, Set<SitePinInst>> deferredRemovals = new HashMap<>();
+
+        EDIFHierCellInst ehciToDelete = netlist.getHierCellInstFromName("processor");
+        List<EDIFHierCellInst> leavesToDelete = netlist.getAllLeafDescendants(ehciToDelete);
+
+        ECOTools.removeCell(design, Collections.singletonList(ehciToDelete), deferredRemovals);
+
+        for (EDIFHierCellInst ehci : leavesToDelete) {
+            String instName = ehci.getFullHierarchicalInstName();
+            // Logical leaf cell not present
+            Assertions.assertNull(netlist.getHierCellInstFromName(instName));
+            // Physical cell not present
+            Assertions.assertNull(design.getCell(instName));
+        }
+
+        // Logical hierarchical cell not present
+        Assertions.assertNull(netlist.getHierCellInstFromName(ehciToDelete.getFullHierarchicalInstName()));
+
+        DesignTools.batchRemoveSitePins(deferredRemovals, true);
+
+        design.writeCheckpoint("/group/zircon2/eddieh/pb.dcp");
+
+        if (FileTools.isVivadoOnPath()) {
+            ReportRouteStatusResult rrs = VivadoTools.reportRouteStatus(design);
+            Assertions.assertEquals(0 /* TODO */, rrs.netsWithRoutingErrors);
+        }
+    }
+
+    @Test
+    public void testRemoveCellLeaf() {
+        Design design = RapidWrightDCP.loadDCP("picoblaze_ooc_X10Y235.dcp");
+        EDIFNetlist netlist = design.getNetlist();
+        Map<Net, Set<SitePinInst>> deferredRemovals = new HashMap<>();
+
+        EDIFHierCellInst ehciToDelete = netlist.getHierCellInstFromName("your_program/ram_4096x8");
+        Assertions.assertTrue(ehciToDelete.getCellType().isLeafCellOrBlackBox());
+
+        ECOTools.removeCell(design, Collections.singletonList(ehciToDelete), deferredRemovals);
+
+        String instName = ehciToDelete.getFullHierarchicalInstName();
+
+        // Logical leaf cell not present
+        Assertions.assertNull(netlist.getHierCellInstFromName(instName));
+        // Physical cell not present
+        Assertions.assertNull(design.getCell(instName));
+
+        DesignTools.batchRemoveSitePins(deferredRemovals, true);
+
+        if (FileTools.isVivadoOnPath()) {
+            ReportRouteStatusResult rrs = VivadoTools.reportRouteStatus(design);
+            Assertions.assertEquals(8, rrs.netsWithRoutingErrors);
+            Assertions.assertEquals(8, rrs.netsWithNoDriver);
+            Assertions.assertEquals(8, rrs.netsWithSomeUnroutedPins);
+        }
+    }
+
+    @Test
+    public void testCreateCell() {
+        Design design = RapidWrightDCP.loadDCP("picoblaze_ooc_X10Y235.dcp");
+        EDIFNetlist netlist = design.getNetlist();
+
+        DesignTools.updatePinsIsRouted(design);
+
+        EDIFCell reference = netlist.getCell("kcpsm6");
+        List<String> instNames = Arrays.asList("processor2", "processor3");
+        ECOTools.createCell(design, reference, instNames);
```

Comment:
It seems `ECOTools.createCell()` returns an object.  Should we capture the return value and also test it?

Reply:
Actually, `ECOTools.createCell()` does not return anything currently. It can be easily enhanced to return an EDIFHierCellInst if that's more meaningful?